File: 7. scrapper/Python/scrapper.py
import requests
from bs4 import BeautifulSoup
import json
from datetime import datetime

# Function to fetch a page
def fetch_page(url):
    try:
        response = requests.get(url)
        response.raise_for_status()  # Raise an error for bad responses
        return BeautifulSoup(response.text, 'html.parser')
    except requests.exceptions.RequestException as e:
        print(f"Error fetching the URL: {url}", e)
        return None

# Function to scrape all pages
def scrape_all_pages():
    base_url = "https://999.md/ro/list/transport/cars?page="
    page = 1
    no_announcements = False
    today = datetime.now()
    formatted_date = f"{today.year}_{today.month:02}_{today.day:02}"
    results_file = f"scrapped_{formatted_date}.json"

    # Initialize the results file
    with open(results_file, 'w') as file:
        file.write("[")

    while not no_announcements:
        url = f"{base_url}{page}"
        soup = fetch_page(url)

        if soup is None:
            print(f"Failed to load page {page}")
            break

        announcements = soup.select(".ads-list-photo.large-photo .ads-list-photo-item")

        if len(announcements) == 0:
            no_announcements = True
            print(f"No announcements found on page {page}. Stopping.")
            break

        page_results = []

        for element in announcements:
            classes = element.get("class", [])
            if "js-booster-inline" in classes or "is-adsense" in classes:
                continue

            title_element = element.select_one(".ads-list-photo-item-title a")
            price_element = element.select_one(".ads-list-photo-item-price-wrapper")

            title = title_element.get_text(strip=True) if title_element else None
            price = price_element.get_text(strip=True).replace("\xa0", " ") if price_element else None
            href = title_element.get("href") if title_element else None
            id = href.split("/").pop() if href else None
            img_element = element.select_one(".ads-list-photo-item-thumb img")
            img_url = img_element.get("src") if img_element else None

            if not title or not id:
                continue

            page_results.append({
                "id": id,
                "title": title,
                "price": price,
                "image": img_url,
            })

        if page_results:
            with open(results_file, 'a') as file:
                data = json.dumps(page_results, indent=4)[1:-1]  # Remove surrounding brackets
                file.write((", " if page > 1 else "") + data)

        print(f"Page {page} processed.")
        page += 1

    # Finalize the results file
    with open(results_file, 'a') as file:
        file.write("]")

    print(f"Results have been saved to {results_file}")

# Start scraping
scrape_all_pages()

File: 7. scrapper/JS/scrapper.js
const axios = require("axios");
const cheerio = require("cheerio");
const fs = require("fs");

async function fetchPage(url) {
  try {
    const { data } = await axios.get(url);
    return cheerio.load(data);
  } catch (error) {
    console.error(`Error fetching the URL: ${url}`, error);
    return null;
  }
}

async function scrapeAllPages() {
  const baseUrl = "https://999.md/ro/list/transport/cars?page=";
  let page = 1;
  let noAnnouncements = false;
  const today = new Date();
  const formattedDate = today.getFullYear() + "_" +
                      String(today.getMonth() + 1).padStart(2, '0') + "_" +
                      String(today.getDate()).padStart(2, '0');
  const resultsFile = `scrapped_${formattedDate}.json`;

  // Initialize the results file
  fs.writeFileSync(resultsFile, "[");

  while (!noAnnouncements) {
    const url = `${baseUrl}${page}`;
    const $ = await fetchPage(url);

    if (!$) {
      console.error(`Failed to load page ${page}`);
      break;
    }

    const announcements = $(".ads-list-photo.large-photo .ads-list-photo-item");

    if (announcements.length === 0) {
      noAnnouncements = true;
      console.log(`No announcements found on page ${page}. Stopping.`);
      break;
    }

    let pageResults = [];

    announcements.each((index, element) => {
      const $element = $(element);
      const classes = $element.attr("class");
      if (
        classes.includes("js-booster-inline") ||
        classes.includes("is-adsense")
      ) {
        return;
      }

      const titlu_element = $element.find(".ads-list-photo-item-title a");
      const pret_element = $element.find(".ads-list-photo-item-price-wrapper");

      const titlu = titlu_element.text().trim();
      const pret = pret_element.text().trim().replace(" ", " ");
      const href = titlu_element.attr("href");
      const id = href ? href.split("/").pop() : null;
      const img_element = $element.find(".ads-list-photo-item-thumb img");
      const img_url = img_element?.attr("src");

      if (!titlu || !id) {
        return;
      }

      pageResults.push({
        id: id,
        title: titlu,
        price: pret,
        image: img_url,
      });
    });

    if (pageResults.length > 0) {
      const data = JSON.stringify(pageResults, null, 4).slice(1, -1); // Remove surrounding brackets
      fs.appendFileSync(resultsFile, (page > 1 ? "," : "") + data);
    }

    console.log(`Page ${page} processed.`);
    page++;
  }

  // Finalize the results file
  fs.appendFileSync(resultsFile, "]");
  console.log("Results have been saved to results.json");
}

// Start scraping
scrapeAllPages();
